File: 9663_N-Queen.cpp
#include <iostream>

using namespace std;

int n, result;
int col[15];

int check(int row);
int nQueen(int index);

int main() {
    cin >> n;
    nQueen(0);
    cout << result << endl;
}

int check(int index) {
    for (int i = 0; i < index; i++) {
        if (col[index] == col[i] || abs(col[index] - col[i]) == index - i) return false;
    }
    return true;
}

int nQueen(int index) {
    if (index == n) {
        result += 1;
    } else {
        for (int i = 0; i < n; i++) {
            col[index] = i;
            if(check(index)) nQueen(index + 1);
        }
    }
    return 0;
}

File: 9663_N-Queen.swift
// 시간 초과

import Foundation

let n = Int(readLine()!)!
var result = 0
var col = [Int](repeating: 0, count: 15)

nQueen(index: 0)
print(result)

func check(index: Int) -> Bool {
    for i in 0..<index {
        if col[index] == col[i] || abs(col[index] - col[i]) == index - i { return false }
    }
    return true
}

func nQueen(index: Int) {
    if index == n {
        result += 1
    } else {
        for i in 0..<n {
            col[index] = i
            if check(index: index) {
                nQueen(index: index + 1)
            }
        }
    }
}
